test/test_utils.py

```diff
@@ -0,0 +1,321 @@
+        shutil.rmtree(DATA_DIRECTORY, onerror=rm_on_error_handler)
+    stop_all_clis()
+    stop_controller()
+    stop_all_test_containers()
+    stop_container(CONTROLLER_TESTS_CONTAINER)
+    remove_all_volumes()
+    if interrupted:
+        sys.exit(1)
+
+
+# Signal handler
+def sig_cleanup(_, __):
+    cleanup(interrupted=True)
+
+
+class DoCleanupContextManager(ContextManager):
+    def __init__(self):
+        self.interrupted = False
+
+    def __exit__(self, *args):
+        cleanup(self.interrupted)
+
+
+def stop_container(container_name: str):
+    if not container_exists(container_name):
+        return
+    execute_command(
+        ['docker',
+         'container',
+         'stop',
+         container_name
+         ],
+        hide_output=True,
+        fail_on_failure=False)
+    execute_command(
+        ['docker',
+         'container',
+         'rm',
+         container_name
+         ],
+        hide_output=True,
+        fail_on_failure=False)
+
+
+def remove_all_volumes():
+    stdout_holder: List[bytes] = []
+    execute_command(
+        [
+            'docker',
+            'volume',
+            'ls',
+            '--quiet',
+            f'--filter=name={VOLUME_PREFIX}'
+        ],
+        stdout_holder=stdout_holder)
+    volumes = str(b''.join(stdout_holder), 'utf-8').splitlines()
+    for volume in volumes:
+        remove_volume(volume)
+
+
+def stop_all_clis():
+    stdout_holder: List[bytes] = []
+    execute_command(
+        ['docker',
+         'container',
+         'ls',
+         '--quiet',
+         '--all',
+         f'--filter=name={CLI_CONTAINER_PREFIX}'
+         ],
+        stdout_holder=stdout_holder)
+    containers = b''.join(stdout_holder)
+    for container in str(containers, 'utf-8').splitlines():
+        stop_container(container)
+
+
+def stop_controller():
+    docker_compose('stop')
+    docker_compose('logs')
+    docker_compose('down', '--volumes')
+
+
+def stop_all_test_containers():
+    stdout_holder = []
+    execute_command(
+        ['docker',
+         'container',
+         'ls',
+         '-a',
+         '--filter=name=plz.execution.id.',
+         '--format={{.ID}}#{{.Image}}'],
+        stdout_holder=stdout_holder,
+        hide_output=True)
+    for l in str(b''.join(stdout_holder), 'utf-8').splitlines():
+        container_id, image = l.split('#', 1)
+        if image.startswith(f'plz/builds:{PLZ_USER}-'):
+            stop_container(container_id)
+
+
+def docker_compose(*args):
+    # Specify the container name, as to avoid the random suffix added by docker
+    # compose
+    env = os.environ.copy()
+    env['CONTROLLER_CONTAINER'] = CONTROLLER_CONTAINER
+    env['DATA_DIRECTORY'] = DATA_DIRECTORY
+    env['REDIS_DATA_DIRECTORY'] = REDIS_DATA_DIRECTORY
+    execute_command(
+        ['docker-compose',
+         f'--project-name={PROJECT_NAME}',
+         f'--file={os.path.join(TEST_DIRECTORY, "docker-compose.yml")}',
+         *args],
+        env=env)
```

Comment:
Gotta break this down. At a glance, I can see the following files extracted:

- *constants.py*
- *docker.py*
- *execution.py* (assuming you can't just get rid of it)
- *log.py*
- *test_context.py* (not "cleanup context; that's the action, not the purpose)